File: ruby/udp/server_read.rb
#prof vou anotar do lado oq faz blz? vou usar ruby pq achei top quando o senhor pesquisou em aula kk

require 'socket' #importa o socket 

# Função pra iniciar o servidor, meio obvio pelo nome, e é bem parecido com python
def start_server(host_read, port_read, host_write, port_write)
    server_read = UDPSocket.new # cria um servidor UDP hahaha, to aprendendo slk
    server_read.bind(host_read, port_read)  # liga o servidor a porta e host

    puts "Server [READ] started at #{host_read}:#{port_read}\nListening now..." # kkkk 'puts', printa o host e porta do servidor

    server_write = UDPSocket.new # cria um servidor UDP pra escrever, o mesmo q ali em cima

    loop do #em ruby é loop do, em python é while True
        data, addr = server_read.recvfrom(1024) # recebe os dados do cliente,igual em python        
        message = data.force_encoding('UTF-8').split(',') # força a codificação dos dados recebidos e separa por vírgula, muda pouca coisa

        data_client = "#{message[0]}, #{message[1]}, #{addr}".strip.force_encoding('UTF-8') # re-encoda os dados recebidos e tira os espaços em branco
        puts "Mensagem do #{data_client}" # printa os dados recebidos do cliente

        server_write.send(data_client, 0, host_write, port_write) # envia os dados para o servidor de escrita
    end
end

if __FILE__ == $0 # ao inves de if name = main é file = $0 no ruby, o resto daqui pra baixo é igual só adicionei um end no final pq no ruby é assim q se fexham 'blocos'
    host_read = 'localhost'
    port_read = 8000

    host_write = 'localhost'
    port_write = 9000

    start_server(host_read, port_read, host_write, port_write)  
end

File: python/udp/server_read.py
import socket

# Função principal do servidor
def start_server(host_read: str, port_read: int, host_write: str, port_write: int):
    server_read = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)  # socket UDP pra receber
    server_read.bind((host_read, port_read))  # conecta na porta/IP pra escutar

    print(f"Server [READ] started at {host_read}:{port_read}\nListening now...")

    server_write = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)  # socket UDP pra enviar

    while True:
        #listening for incoming data
        data, addr = server_read.recvfrom(1024)  # recebe dados do cliente
        message = data.decode('utf-8').split(',')  # transforma de bytes pra string
        
        #data_client is string
        data_client = f'{message[0]}, {message[1]}, {addr}'.strip().encode('utf-8')
        print(data_client)
        
        #send data to the server_read
        server_write.sendto(data_client, (host_write, port_write))


# Configurações e chamada da função principal
if __name__ == '__main__':
    HOST_READ = 'localhost'   # onde vai escutar
    PORT_READ = 8000          # porta de entrada

    HOST_WRITE = 'localhost'  # pra onde vai mandar os dados
    PORT_WRITE = 9000         # porta de saída

    start_server(HOST_READ, PORT_READ, HOST_WRITE, PORT_WRITE)  # inicia o servidor
